pkg/lib/cockpit.d.ts

```diff
@@ -24,6 +24,11 @@ declare module 'cockpit' {
         problem: string;
         message: string;
         toString(): string;
+
+        // when using the "host" option
```

Comment:
BasicError is defined at line 1040 in `cockpit.js` and contains only three attributes.  These definitely aren't among them.

This looks more like you're dealing with some JSON blob that was a close or failed init control message?

Reply:
OK, then we need to rename this as something else (`CockpitError`?), or I'll just ignore the types in cockpit-connect-ssh for now. In JS it doesn't make much sense to have more than one error class, as both Promise.catch() and `catch()` can only appear once (unlike e.g. `except FooBarError:` in Python). This doesn't really go well together with typing.

I suppose that's also why tsc has this special-cased?

> Catch clause variable type annotation must be 'any' or 'unknown' if specified.

There's precedent:
```
pkg/base1/test-fsinfo.ts:    } catch (exc: any) { // eslint-disable-line @typescript-eslint/no-explicit-any
```

I'll drop that bit, we can discuss it in the next PR.

Reply:
fwiw, in my "new channel API" branch I defined an `ErrorLike` which is anything that has a `string` `problem` and `message` property.  All of our Error types fit that description.  Then I modified our DeferredPromise API to always return that on the `.fail()` path and modified `cockpit.message()` to accept it.

I think something like that could be helpful.  Or just `: any` everywhere, honestly.  Exception handling in JavaScript is a typing train wreck.  This is one place where I wouldn't mind a bit of TypeScript sugar, so you could write something like:

```ts
try {
    ...
} catch (exc: BasicError) {
    console.log(exc.message);
}
```

and have that emit the following JS:

```js
try {
    ...
} catch (exc) {
    if (exc instanceof BasicError) {
        console.log(exc.message);
    } else {
        throw exc;
    }
}
```

I guess we won't get that soon, though.  TS are very very conservative about adding these sorts of "language extending" features.

Reply:
Yup.  https://github.com/microsoft/TypeScript/issues/14191

> This is purely a runtime feature and should be pushed into the JS standard if it's sufficiently useful.


So I guess what I'm saying is: maybe just write it out the long way?  It's ugly, but if that becomes the sort of idiom that we want to use, then ... let's do that?  Eventually we'll start reading it as if it were Python...

And we don't have to do this using nominal RTTI: we can also do it structurally, dynamically.

We can also define type guards for things like "check if this object has problem and message codes hanging on it".  Sort of something like:

```ts

interface ErrorLike {
  problem: string;
  message: string;
}

function is_errror_like(value: unknown): value is ErrorLike {
  // ...
}

// ...later...

try {
  ...
} catch (exc) {
  if (is_error_like(exc)) {
    console.log(exc.message);
  } else {
    throw exc;
  }
}
```

You could almost format/read that as:

```ts
try {
  ...
} catch (exc) { if (is_error_like(exc)) {
  console.log(exc.message);
} else { throw exc; } }
```

but let's not ;)